### PR TITLE
register python exceptions and update exceptions handling

### PR DESCRIPTION
**Pull request by askhade**
Signed-off-by: Ashwini Khade <askhade@microsoft.com>

### CODE REVIEW

File: onnx/shape_inference/implementation.cc [hunk 1]
```diff
@@ -292,11 +292,10 @@ static void InferShapesImpl(
               function_opset_imports[opset_import.domain()] = static_cast<int>(opset_import.version());
             } else {
               if (it->second != opset_import.version()) {
-                throw ONNX_NAMESPACE::InferenceError(
-                    "ONNX models do not support multiple opset version imports for a domain. Function " +
-                            schema->Name() + " imports opset version " + std::to_string(opset_import.version()) +
-                            " for domain " + (opset_import.domain().empty() ? "ai.onnx" : opset_import.domain()) +
-                            " where as the model imports opset version " + std::to_string(it->second));
+                fail_shape_inference("ONNX models don't support multiple opset version imports for a domain. Function ",
```
Comment (jcwchen) [inline onnx/shape_inference/implementation.cc:295]: Is it a "shape" inference? I thought using ONNX_NAMESPACE::InferenceError is more proper

**Review by jcwchen — COMMENTED** [review]
Others LGTM. Thank you so much for organizing exceptions.

File: onnx/cpp2py_export.cc [hunk 2]
```diff
@@ -324,6 +327,8 @@ PYBIND11_MODULE(onnx_cpp2py_export, onnx_cpp2py_export) {
   // Submodule `shape_inference`
   auto shape_inference = onnx_cpp2py_export.def_submodule("shape_inference");
   shape_inference.doc() = "Shape Inference submodule";
+  py::register_exception<InferenceError>(shape_inference, "InferenceError");
```
Comment (gramalingam) [inline onnx/cpp2py_export.cc:330]: Is this intended to cover both shape-inference errors as well as type-errors? Wondering why module is called "shape inference", while the exception is plain "InferenceError".

File: onnx/shape_inference/implementation.cc (hunk 1 above)
Comment (askhade) [reply onnx/shape_inference/implementation.cc:295]: fail_shape_inference throws a InferenceError... not sure what you mean

File: onnx/cpp2py_export.cc (hunk 2 above)
Comment (askhade) [reply onnx/cpp2py_export.cc:330]: yes both module and exception class are intended to handle both type and shape inference.... I am simply registering the exception here...

File: onnx/shape_inference/implementation.cc (hunk 1 above)
Comment (jcwchen) [reply onnx/shape_inference/implementation.cc:299]: Yes eventually they will throw the same error. Just thought this one is actually not a "shape" inference error so directly using `fail_shape_inference` looks a little weird to me.

**Review by jcwchen — APPROVED** [review]

File: onnx/cpp2py_export.cc [hunk 3]
```diff
@@ -334,19 +334,19 @@ PYBIND11_MODULE(onnx_cpp2py_export, onnx_cpp2py_export) {
   py::register_exception<InferenceError>(shape_inference, "InferenceError");
 
 
-  shape_inference.def("infer_shapes", [](const py::bytes& bytes, bool check_type) {
+  shape_inference.def("infer_shapes", [](const py::bytes& bytes, bool check_type, bool strict_mode) {
```
Comment (gramalingam) [inline onnx/cpp2py_export.cc:337]: I wonder if making strict_mode into an integer (called mode, restricted to values 0 and 1 currently) will give us more flexibility in the future. For example, may be we could treat it as a compatibility_mode indicator indicating that we desire the behavior in version 1.7 or 1.8, etc. In the future, we could potentially end up with more than 2 modes from a compatibility-mode perspective.

Comment (askhade) [reply onnx/cpp2py_export.cc:337]: I am leaning towards still keeping the strict_mode as true\false in python and change the one in c++ interface to int... when we add more granularity we can still keep the python api for backward compatibility and add a new api in python for the new approach

**Review by gramalingam — APPROVED** [review]

**Pull request merged**